input/input.go

```diff
@@ -0,0 +1,252 @@
+	RepositoryType string `yaml:"repo_type" json:"repo_type"`
+	// RepositoryName represents the name of the repository. <REQUIRED>
+	RepositoryName string `yaml:"repo_name" json:"repo_name"`
+	// RepositoryConfig defines reposiroty config. <REQUIRED>
+	RepositoryConfig `yaml:"repo_config" json:"repo_config"`
+}
+
+// RepositoryConfig represents repostory configurations.
+type RepositoryConfig struct {
+	// RepositoryURL is the url for the repository. <REQUIRED>
+	RepositoryURL string `yaml:"repo_url" json:"repo_url"`
+	// RepositoryCredentials represents repository credentials. <REQUIRED>
+	RepositoryCredentials `yaml:"credentials" json:"credentials"`
+	//Branches represents branches that needs to be monitored. <OPTIONAL>
+	Branches []string `yaml:"branches" json:"branches"`
+}
+
+// RepositoryCredentials consists of credential needed to authenticate with repository.
+type RepositoryCredentials struct {
+	//Either username or email is <REQUIRED>
+	//Username of the repository.
+	Username string `yaml:"username" json:"username"`
+	//Email of the repository.
+	Email string `yaml:"email" json:"email"`
+	//AccessToken for accessing the github's APIs. <REQUIRED>
+	AccessToken string `yaml:"access_token" json:"access_token"`
+}
+
+// Output represents the target where data will be sent.
+type Output struct {
+	//TargetName consists of the target names to which auditjob data needs to be sent. <REQUIRED>
+	TargetName []string `yaml:"target_name" json:"target_name"`
+}
+
+// Tags are additional information that can be added in tags fields of output documents.
+type Tags map[string]string
+
+// Target represents the target where data will be published.
+type Target struct {
+	//Name of the target. <REQUIRED>
+	Name string `yaml:"name" json:"name"`
+	//Type of the target , possible values (elasticsearch , kafka-rest etc). <REQUIRED>
+	Type string `yaml:"type" json:"type"`
+	//TargetConfig consist of the target configurations.
+	TargetConfig `yaml:"config" json:"config"`
+}
+
+//TargetConfig is key value target configurations.
+type TargetConfig map[string]string
+
+// InitConfig reads the config file and populates Config struct with relevant data. It returns error in case of file read error based on ReadConfigFile().
+func InitConfig(configPath string) (Config, error) {
+	log.Infof("reading config file located at %s", configPath)
+	var conf Config
+	data, err := ReadConfigFile(configPath)
+	if err != nil {
+		return Config{}, err
+	}
+	if err := yaml.Unmarshal(data, &conf); err != nil {
+		log.Errorf("error[%v] in unmarshaling config file to Config struct", err)
+		return Config{}, err
+	}
+	// validating config
+	log.Infof("validating fields in config file located at %s", configPath)
+	err = conf.validate()
+	if err != nil {
+		log.Errorf("error[%v] in validating config file", err)
+		return Config{}, err
+	}
+	// adding default values for missing fields
+	log.Infof("populating default field values in github-audit config")
+	conf.populateDefaultValues()
```

Comment:
Default values should be populated before validation right?

Reply:
yes logical error fixed